File: Recursion/RecursionJava/src/SortingInArrays/MergeSort.java
package SortingInArrays;

import java.util.ArrayList;
import java.util.Arrays;

public class MergeSort {
    public static void main(String[] args) {
        int []arr = {8,7,6,5,4,3,2,1};

        mergeSortInPlace(arr,0, arr.length);
        System.out.println(Arrays.toString(arr));

    }
    public static int [] sorting(int []arr ){
        if (arr.length == 1 ){
            return arr;
        }
        int mid = arr.length / 2;
        int []left = sorting(Arrays.copyOfRange(arr,0,mid));
        int []right = sorting(Arrays.copyOfRange(arr,mid,arr.length));

        return merge(left,right);
    }

    public static int[] merge(int []first , int []second){
        int []ans = new int[first.length+ second.length];
        int i = 0 ;
        int j  =  0 ;
        int k = 0;

        while(i< first.length && j < second.length){
            if (first[i] < second[j]){
                ans[k] = first[i];
                i++;
            }else{
                ans[k] = second[j];
                j++;
            }
            k++;
        }
        while(i< first.length){
            ans[k] = first[i];
            i++;
            k++;
        }
        while(j< second.length){
            ans[k] = second[j];
            j++;
            k++;
        }
        return ans;
    }



    public static void mergeSortInPlace(int[]arr , int start ,int end){
        if (end - start == 1){
            return;
        }
        int mid = start + (end -  start ) / 2;
        mergeSortInPlace(arr,start,mid);
        mergeSortInPlace(arr,mid,end);

        mergeInPlace(arr,start,mid,end);
    }

    public static  void mergeInPlace(int []arr , int start ,int mid , int end ){
        int [] ans = new int[end - start];
        int i = start ;
        int j = mid ;
        int k = 0;
        
        while(i< mid && j < end){
            if (arr[i] < arr[j]){
                ans[k] = arr[i];
                i++;
            }else{
                ans[k] = arr[j];
                j++;
            }
            k++;
        }
        while(i< mid){
            ans[k] = arr[i];
            i++;
            k++;
        }
        while(j< end){
            ans[k] = arr[j];
            j++;
            k++;
        }
        for (int l = 0; l < ans.length ; l++) {
            arr[start+l] = ans[l];
        }

    }


}

File: Recursion/RecursionC++/mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int *arr ,int start ,int end){
	int mid  = (start+ end) / 2;

	int len1 = mid - start + 1;
	int len2 = end - mid ;
	int *first = new int[len1] ;
	int *second =  new int[len2] ;
	int k = start;

	for(int i = 0 ; i < len1 ;i++){
		first[i] = arr[k++];
	}
	k = mid + 1;
	for(int i = 0 ; i < len2 ;i++){
		second[i] = arr[k++];
	}

	int index1  = 0 ;
	int  index2 =  0 ;
	k = start ;

	while(index1 < len1 && index2 < len2){
		if(first[index1] < second[index2]){
			arr[k] = first[index1] ;
			index1++;
		}
		else if(first[index1] > second[index2]){
			arr[k] = second[index2] ;
			index2++;
		}
		k++;
	}
	while( index1 < len1){
		arr[k] = first[index1] ;
		index1++;
		k++;
	}
	while( index2 < len2){
		arr[k] = second[index2] ;
		index2++;
		k++;
	}

}

void mergeSort(int *arr , int start ,int end){
	if(start >= end){
		return;
	}
	int mid = (start + end)/2 ;
	mergeSort(arr , start , mid);
	mergeSort(arr , mid+1 ,end );
	merge(arr , start , end);
}

int main(){
	int  n;
	cin >> n ;
	int arr[n];
	for(int i = 0 ; i < n ; i++){
		cin >> arr[i] ;
	}
	mergeSort(arr , 0 , n);
	for(int i = 0 ; i < n ; i++){
		cout << arr[i] << " " ;
	}
}
